### ISSUE TITLE
ResolveProcess occasionally results in an NPE

### ISSUE THREAD

**Issue by rotty3000** [bndtools/bnd#1697]
If you've ever seen an NPE at (felix resolver) Candidates:917 it's because the while most of the resolver supports returning null from findProviders, fragment handling does not. In addition, the javadoc for the findProviders method does not state that it can return null. 

The solution is to never return null.

**Comment by rotty3000** [comment]
Here's the stacktrace in question:

```
* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':web.sample:resolve.web.sample'.
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:69)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:46)
	at org.gradle.api.internal.tasks.execution.PostExecutionAnalysisTaskExecuter.execute(PostExecutionAnalysisTaskExecuter.java:35)
	at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:68)
	at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:58)
	at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:52)
	at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
	at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:53)
	at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
	at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:203)
	at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:185)
	at org.gradle.execution.taskgraph.AbstractTaskPlanExecutor$TaskExecutorWorker.processTask(AbstractTaskPlanExecutor.java:66)
	at org.gradle.execution.taskgraph.AbstractTaskPlanExecutor$TaskExecutorWorker.run(AbstractTaskPlanExecutor.java:50)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor.process(DefaultTaskPlanExecutor.java:25)
	at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter.execute(DefaultTaskGraphExecuter.java:110)
	at org.gradle.execution.SelectedTaskExecutionAction.execute(SelectedTaskExecutionAction.java:37)
	at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:37)
	at org.gradle.execution.DefaultBuildExecuter.access$000(DefaultBuildExecuter.java:23)
	at org.gradle.execution.DefaultBuildExecuter$1.proceed(DefaultBuildExecuter.java:43)
	at org.gradle.execution.DryRunBuildExecutionAction.execute(DryRunBuildExecutionAction.java:32)
	at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:37)
	at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:30)
	at org.gradle.initialization.DefaultGradleLauncher$4.run(DefaultGradleLauncher.java:154)
	at org.gradle.internal.Factories$1.create(Factories.java:22)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:90)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:52)
	at org.gradle.initialization.DefaultGradleLauncher.doBuildStages(DefaultGradleLauncher.java:151)
	at org.gradle.initialization.DefaultGradleLauncher.access$200(DefaultGradleLauncher.java:32)
	at org.gradle.initialization.DefaultGradleLauncher$1.create(DefaultGradleLauncher.java:99)
	at org.gradle.initialization.DefaultGradleLauncher$1.create(DefaultGradleLauncher.java:93)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:90)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:62)
	at org.gradle.initialization.DefaultGradleLauncher.doBuild(DefaultGradleLauncher.java:93)
	at org.gradle.initialization.DefaultGradleLauncher.run(DefaultGradleLauncher.java:82)
	at org.gradle.launcher.exec.InProcessBuildActionExecuter$DefaultBuildController.run(InProcessBuildActionExecuter.java:94)
	at org.gradle.tooling.internal.provider.ExecuteBuildActionRunner.run(ExecuteBuildActionRunner.java:28)
	at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
	at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:43)
	at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:28)
	at org.gradle.launcher.exec.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:81)
	at org.gradle.launcher.exec.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:46)
	at org.gradle.launcher.exec.DaemonUsageSuggestingBuildActionExecuter.execute(DaemonUsageSuggestingBuildActionExecuter.java:51)
	at org.gradle.launcher.exec.DaemonUsageSuggestingBuildActionExecuter.execute(DaemonUsageSuggestingBuildActionExecuter.java:28)
	at org.gradle.launcher.cli.RunBuildAction.run(RunBuildAction.java:43)
	at org.gradle.internal.Actions$RunnableActionAdapter.execute(Actions.java:173)
	at org.gradle.launcher.cli.CommandLineActionFactory$ParseAndBuildAction.execute(CommandLineActionFactory.java:241)
	at org.gradle.launcher.cli.CommandLineActionFactory$ParseAndBuildAction.execute(CommandLineActionFactory.java:214)
	at org.gradle.launcher.cli.JavaRuntimeValidationAction.execute(JavaRuntimeValidationAction.java:35)
	at org.gradle.launcher.cli.JavaRuntimeValidationAction.execute(JavaRuntimeValidationAction.java:24)
	at org.gradle.launcher.cli.CommandLineActionFactory$WithLogging.execute(CommandLineActionFactory.java:207)
	at org.gradle.launcher.cli.CommandLineActionFactory$WithLogging.execute(CommandLineActionFactory.java:169)
	at org.gradle.launcher.cli.ExceptionReportingAction.execute(ExceptionReportingAction.java:33)
	at org.gradle.launcher.cli.ExceptionReportingAction.execute(ExceptionReportingAction.java:22)
	at org.gradle.launcher.Main.doAction(Main.java:33)
	at org.gradle.launcher.bootstrap.EntryPoint.run(EntryPoint.java:45)
	at org.gradle.launcher.bootstrap.ProcessBootstrap.runNoExit(ProcessBootstrap.java:55)
	at org.gradle.launcher.bootstrap.ProcessBootstrap.run(ProcessBootstrap.java:36)
	at org.gradle.launcher.GradleMain.main(GradleMain.java:23)
	at org.gradle.wrapper.BootstrapMainStarter.start(BootstrapMainStarter.java:30)
	at org.gradle.wrapper.WrapperExecutor.execute(WrapperExecutor.java:129)
	at org.gradle.wrapper.GradleWrapperMain.main(GradleWrapperMain.java:61)
Caused by: java.lang.NullPointerException
	at org.apache.felix.resolver.util.CopyOnWriteList.<init>(CopyOnWriteList.java:42)
	at org.apache.felix.resolver.util.ShadowList.<init>(ShadowList.java:31)
	at org.apache.felix.resolver.Candidates.prepare(Candidates.java:917)
	at org.apache.felix.resolver.ResolverImpl.resolve(ResolverImpl.java:232)
	at org.apache.felix.resolver.ResolverImpl.resolve(ResolverImpl.java:179)
	at biz.aQute.resolve.ResolveProcess.resolveRequired(ResolveProcess.java:72)
	at biz.aQute.resolve.ProjectResolver.resolve(ProjectResolver.java:96)
	at biz.aQute.resolve.ProjectResolver$resolve.call(Unknown Source)
	at aQute.bnd.gradle.BndPlugin$_apply_closure1_closure25_closure71_closure72.doCall(BndPlugin.groovy:445)
	at org.gradle.api.internal.AbstractTask$ClosureTaskAction.execute(AbstractTask.java:554)
	at org.gradle.api.internal.AbstractTask$ClosureTaskAction.execute(AbstractTask.java:535)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:80)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:61)
	... 60 more

```

**Comment by bjhargrave** [comment]
This is not the BndPlugin in the master branch. This must be a fork of yours.

It is possible this is a bug in the Felix Resolver coder. Have you tried this with the latest Felix Resolver? It is also possible the inputs to the resolver are bad, perhaps there is a bug in your code?

**Comment by bjhargrave** [comment]
> If you've ever seen an NPE at (felix resolver) Candidates:917 it's because the while most of the resolver supports returning null from findProviders, fragment handling does not. In addition, the javadoc for the findProviders method does not state that it can return null.
> 
> The solution is to never return null.

Are you saying that the ResolveContext.findProviders method returns null to the resolver instead of List? Doing so would violate the OSGi spec. If the resolver handles this in some cases, then it is being generous as it should just demand a List returned from findProviders.

**Comment by tjwatson** [comment]
Took a quick look at the resolver code for 1.8.0.  AFAICT it only calls ResolveContext.findProviders in one place.  And that place will fail with an NPE pretty quick (much earlier than the one reported here):

http://svn.apache.org/viewvc/felix/trunk/resolver/src/main/java/org/apache/felix/resolver/Candidates.java?view=markup&pathrev=1696846#l227

            List<Capability> candidates = rc.findProviders(requirement);
            LinkedList<Resource> newToPopulate = new LinkedList<Resource>();
            ResolutionError thrown = processCandidates(rc, newToPopulate, requirement, candidates);
             if (candidates.isEmpty() && !Util.isOptional(requirement))

The call to processCandidates immediately tries to get an Iterator from the List which would NPE.
I suspect something else is going on.  Perhaps some optional requirement?  But that doesn't really make sense either since this part where the NPE is going over the m_dependentMap which maps a capability back to the requirements that match it, then the requirement is used to look up all the matching capability (List) for that requirement.  There should be at least one here since the requirement was int the value for the m_depenentMap.

**Comment by bjhargrave** [comment]
> Are you saying that the ResolveContext.findProviders method returns null to the resolver instead of List?

Like @seanbright, I don't see how findProviders0 could return null. And based upon @tjwatson's comment, I think this must be some issue other than findProviders returning null.

**Comment by seanbright** [comment]
The relevant path in the resolver code has changed significantly since 1.8.0 was released. @tjwatson, would your changes in r1746343 (FELIX-5254) have potentially "fixed" this particular problem, or do you think it would just fail in a different way?

**Comment by seanbright** [comment]
So from a resolver standpoint, the map that the `null` is coming out of is `m_candidateMap` in `Candidates` and the only values that are being `put()` into that map are the result of calling `new SomeKindOfList()...` (so non-`null`).

Given that and the inability to reliable reproduce this exception, my guess is that there is a thread safety issue in the resolver code.

**Comment by bjhargrave** [comment]
> Given that and the inability to reliable reproduce this exception, my guess is that there is a thread safety issue in the resolver code.

Perhaps we should use the resolver in single threaded mode?

**Comment by seanbright** [comment]
Maybe we ask @rotty3000 to try that with his test workspace for a while and see what happens. I would hate to given up the speed benefits of the multi-threaded resolver for the majority of cases where it doesn't cause a problem.

**Comment by bjhargrave** [comment]
> Perhaps we should use the resolver in single threaded mode?

It is already being used in single threaded mode :-) 
<https://github.com/bndtools/bnd/blob/master/biz.aQute.resolve/src/biz/aQute/resolve/ProjectResolver.java#L81>

**Comment by bjhargrave** [comment]
I made a PR to use the Processor's executor for all resolve operations.

https://github.com/bndtools/bnd/pull/1700

**Comment by seanbright** [comment]
> It is already being used in single threaded mode :-)

well that's embarassing

**Comment by tjwatson** [comment]
Make sure you use your own executer with the resolver.  The built in one creates and destroys one every resolve.  I'm also not sure the parallel work the resolver does helps much.  It would be better if it tried each solution in parallel.   But that is not what it does. 

My guess is there is some randomness to the order thing are done in the resolver.  A certain order is exposing this bug.  My guess is the key being used here is a wrapper when we expect it to be the original req.

I have no reason to believe that my recent changes would fix this. But it would be good to try.

**Comment by bjhargrave** [comment]
I have a PR <https://github.com/bndtools/bnd/pull/1700> to change to use an executor for the resolver.

I also have another PR <https://github.com/bndtools/bnd/pull/1701> which changes the Cap/Req/Res impls to return the same objects each call. Before it always created a new unmodifiable object per call. I don't think this is a big thing, but may help.

**Comment by rotty3000** [comment]
Hey all, sorry, but for some reason my github account was set to ignore alerts from this project... (maybe due to my vacation responder. better check all my other projects...)

**Comment by rotty3000** [comment]
@bjhargrave the code path is simply from having added the resolve task to the gradle plugin. Trying to get some of these exceptions from within bndtools is pointless, so I normally try to debug from a gradle execution. Either way, you will get this NPE in the resolve wizard of bndtools, however, you won't have any details since all you get is the e.getMessage() which is pretty useless for NPEs to begin with.

**Comment by seanbright** [comment]
@rotty3000 you don't happen to have a shareable workspace that (eventually) causes the problem, do you?

**Comment by rotty3000** [comment]
Here's a simple project that has demonstrated the issue. (it's the whole workspace zipped)
[bnd.issue.1706.zip](https://github.com/bndtools/bnd/files/480654/bnd.issue.1706.zip)

**Comment by tjwatson** [comment]
> @tjwatson What version of the resolver jar are these in?
@bjhargrave It does not appear the felix project has done a new release for the resolver that includes the latest.  You would have to build from trunk to test.  I have not done a release out of the felix project myself and I will not have time to figure that out for a while.  Not sure if someone else more familiar with felix could do it.

**Comment by bjhargrave** [comment]
> @tjwatson What version of the resolver jar are these in?
> @bjhargrave It does not appear the felix project has done a new release for the resolver that includes the latest. You would have to build from trunk to test. I have not done a release out of the felix project myself and I will not have time to figure that out for a while. Not sure if someone else more familiar with felix could do it.

@karlpauls I think you made the 1.8.0 version of the resolver jar you can download from <http://felix.apache.org/downloads.cgi>. Is there a plan to release a newer version?

**Comment by karlpauls** [comment]
I can do you a new release of the resolver.

**Comment by karlpauls** [comment]
FWIW, we released org.apache.felix.resolver-1.10.0 as part of the Framework 5.6.0 release.

**Comment by bjhargrave** [comment]
@rotty3000 Perhaps you can test with the updated felix resolver to see if the NPE is still an issue?

**Comment by rotty3000** [comment]
will do!

**Comment by rotty3000** [comment]
There still appears to be an issue.

**Comment by rotty3000** [comment]
I'm fairly certain this was fixed in 62cf992

please reopen if you find it's still happening.
